### ISSUE TITLE
UI for password encrypted message functionality

### ISSUE THREAD

**Issue by DenBond7** [FlowCrypt/flowcrypt-android#1620]
See https://github.com/FlowCrypt/flowcrypt-ios/issues/1221

**Comment by DenBond7** [comment]
Actually, it's an interesting issue :). It seems we have a not simple logic here.

**Comment by DenBond7** [comment]
Before starting to work on this issue need to understand what exactly we do here.

As I understand we have the following:

1. During creating a new encrypted message we can have recipients that don't have public keys.
2. We show a hint for a user: you can import missing public keys or protect the current message with a pass phrase for recipients with missing public keys.
3. After that the user should provide missing public keys or a pass phrase that will be used for a password encrypted message.
4. Let's imagine the user has provided a pass phrase. What do we do next?
5. It seems we should create an encrypted message that will be sent as attachment, do all things from #1615, receive some URL and send a message with text 

```
user@flowcrypt.test has sent you a password-encrypted email Click here to Open Message

Alternatively copy and paste the following link: https://flowcrypt.com/HoHoHoHo
```
and attachments(`noname`, `encryped.asc`)

Just a simplified version what I see.

@tomholub Please correct me if needed

**Comment by IvanPizhenko** [comment]
We also should take into account that enterprise customers may have their own FES.

**Comment by DenBond7** [comment]
> We also should take into account that enterprise customers may have their own FES instance.

Yes, it relates to #1621 and #1622

**Comment by DenBond7** [comment]
as an example to discuss

![image](https://user-images.githubusercontent.com/2863246/146935024-47ff256d-9bb5-4f52-8132-77ac1986ecb3.png)

**Comment by tomholub** [comment]
I wanted to do this on one platform first (iOS in this case) so that we can see it in action and settle the exact UI before doing it on a second platform (Android). That's why I tried to assign you non-ui tasks here and assign Roman UI tasks there, so I don't have to be potentially wrong twice per platform for all tasks (because I'm not too sure until I see it).

**Comment by tomholub** [comment]
> We show a hint for a user: you can import missing public keys or protect the current message with a pass phrase for recipients with missing public keys.

we call it a password in this case to differentiate, though it's true it doesn't sound that different

**Comment by tomholub** [comment]
> Let's imagine the user has provided a pass phrase. What do we do next?
> It seems we should create an encrypted message that will be sent as attachment, do all things from prepare API model / client class for sending through web portal #1615, receive some URL and send a message with text

This is outside of scope of this issue, so we don't have to discuss it in detail yet. But basically

1) call FES to get a reply token
2) encode the reply token into the message & encrypt a message with both public keys (my own + recipients if any) and the provided password
3) upload the message to FES
4) format the message with the sentence + link + encrypted data in attachment
5) send message formatted this way through gmail

**Comment by tomholub** [comment]
This is pretty good. The only problem is, it only shows after user already tapped send which is a bit too late. It would be better to have some banner.

Also:
 - the password option should be first, when available
 - switch to standard email should be removed when FES available

![image](https://user-images.githubusercontent.com/6306961/146998104-b72a70a5-f5a5-4af8-9081-777555d58599.png)

**Comment by tomholub** [comment]
@sosnovsky you can have a look at the sketches above in case it helps you with https://github.com/FlowCrypt/flowcrypt-ios/issues/1221

**Comment by sosnovsky** [comment]
> @sosnovsky you can have a look at the sketches above in case it helps you with [FlowCrypt/flowcrypt-ios#1221](https://github.com/FlowCrypt/flowcrypt-ios/issues/1221)

Looks good, I currently have more basic UI implementation:

![Simulator Screen Shot - iPhone 13 - 2021-12-21 at 23 14 59](https://user-images.githubusercontent.com/1062093/146998761-84392979-8200-406d-bd84-352552a48114.png)

![Simulator Screen Shot - iPhone 13 - 2021-12-21 at 23 15 04](https://user-images.githubusercontent.com/1062093/146998756-763a3a07-2947-4da0-8071-1dd1b7174fe3.png)

![Simulator Screen Shot - iPhone 13 - 2021-12-21 at 23 15 15](https://user-images.githubusercontent.com/1062093/146998751-94f22565-bfb6-4d05-bccf-dc84f1fef51f.png)

**Comment by tomholub** [comment]
I think the screenshots on both platforms are appropriate (for the level of implementation / features already available for that platform). So it looks on a good track.

I think on iOS, above the password, we could add another paragraph of text about the user being responsible to share the password. Per the Android screenshot "You are responsible ..."

**Comment by DenBond7** [comment]
@tomholub 
> I wanted to do this on one platform first (iOS in this case) so that we can see it in action and settle the exact UI before doing it on a second platform (Android).

I've provided only UI (after checking the web version.). Just sketches without logic. I have not added logic yet :)

> The only problem is, it only shows after user already tapped send which is a bit too late. 

It depends on our needs. For example, we can show `floatActionButton` when we found at least one recipient without a public key. Similar to the web version does.

**Comment by tomholub** [comment]
Yes. And do UI is not bad. So you can use this.

**Comment by tomholub** [comment]
The float sounds good.

**Comment by tomholub** [comment]
@DenBond7 does Android have any database entry that indicates if FES is available for that user?

**Comment by DenBond7** [comment]
We store OrgRules. If it contains such info (some entry that indicates FES) my answer - yes.

**Comment by DenBond7** [comment]
@tomholub Can we fetch `clientConfiguration` when `FES` doesn't work?

**Comment by tomholub** [comment]
This password-protected message functionality will only be available when FES is used, not when flowcrypt.com/api is used.

Therefore we'll need a boolean in the storage. But not for this specific UI issue, because of this:

> For now, if domain of active user is not flowcrypt.com, keep current behavior (because this new functionality will be work-in-progress for a while, we don't want to allow it for external users yet).

Here I'm referring to the sender. Since flowcrypt.com users have FES, we don't have to worry about detecting if FES is used or not.

Once all of the functionality for password-protected messages is ready and merged, then we'll change this `flowcrypt.com` restriction to become "domain uses FES" restriction. I'll file a separate issue for that.

**Comment by DenBond7** [comment]
@tomholub It seems https://github.com/FlowCrypt/flowcrypt-ios/issues/1221 is already closed.

>I wanted to do this on one platform first (iOS in this case) so that we can see it in action and settle the exact UI before doing it on a second platform (Android). That's why I tried to assign you non-ui tasks here and assign Roman UI tasks there, so I don't have to be potentially wrong twice per platform for all tasks (because I'm not too sure until I see it).

I can to come back to this issue. How can I test the last changes for iOS? Would you like to share any instruction for UI for this issue?

**Comment by tomholub** [comment]
Your UI example, with my comments here https://github.com/FlowCrypt/flowcrypt-android/issues/1620#issuecomment-999096915 looks good to me for Android.

**Comment by tomholub** [comment]
I'll build you the iOS app so you can test it too.

**Comment by DenBond7** [comment]
@tomholub I think UI is done. Please look at the video. Maybe you would like to change something

https://user-images.githubusercontent.com/2863246/149094039-18d3973d-b741-43d7-885c-fd12224d91de.mp4

**Comment by tomholub** [comment]
This is good. Please, for now, hide the strength evaluation - there will be different rules for evaluating strength of passwords (as opposed to pass phrases). Else this looks great.
